src/js/tracing/reactnativenavigation.ts

```diff
@@ -0,0 +1,206 @@
+import {
+  RoutingInstrumentation,
+  TransactionCreator,
+} from "./routingInstrumentation";
+import { getBlankTransactionContext } from "./utils";
+
+interface ReactNativeNavigationOptions {
+  routeChangeTimeoutMs: number;
+}
+
+const defaultOptions: ReactNativeNavigationOptions = {
+  routeChangeTimeoutMs: 1000,
+};
+
+interface ComponentEvent {
+  componentId: string;
+}
+
+type ComponentType =
+  | "Component"
+  | "TopBarTitle"
+  | "TopBarBackground"
+  | "TopBarButton";
+
+export interface ComponentWillAppearEvent extends ComponentEvent {
+  componentName: string;
+  passProps?: Record<string | number | symbol, unknown>;
+  componentType: ComponentType;
+}
+
+export interface EventSubscription {
+  remove(): void;
+}
+
+export interface EventsRegistry {
+  registerComponentWillAppearListener(
+    callback: (event: ComponentWillAppearEvent) => void
+  ): EmitterSubscription;
+  registerCommandListener(
+    callback: (name: string, params: unknown) => void
+  ): EventSubscription;
+}
+
+export interface NavigationDelegate {
+  events: () => EventsRegistry;
+}
+
+/**
+ * Instrumentation for React Native Navigation. See docs or sample app for usage.
+ *
+ * How this works:
+ * - `_onCommand` is called every time a commands happens and sets an IdleTransaction on the scope without any route context.
+ * - `_onComponentWillAppear` is then called AFTER the state change happens due to a dispatch and sets the route context onto the active transaction.
+ * - If `_onComponentWillAppear` isn't called within `options.routeChangeTimeoutMs` of the dispatch, then the transaction is not sampled and finished.
+ */
+export class ReactNativeNavigationInstrumentation extends RoutingInstrumentation {
+  public static instrumentationName: string = "react-native-navigation";
+
+  private _navigation: NavigationDelegate;
+  private _options: ReactNativeNavigationOptions;
+
+  private _prevComponentEvent: ComponentWillAppearEvent | null = null;
+
+  private _latestTransaction?: TransactionType;
+  private _recentComponentIds: string[] = [];
+  private _stateChangeTimeout?: number | undefined;
+
+  public constructor(
+    /** The react native navigation `NavigationDelegate`. This is usually the import named `Navigation`. */
+    navigation: NavigationDelegate,
+    options: Partial<ReactNativeNavigationOptions> = {}
+  ) {
+    super();
+
+    this._navigation = navigation;
+
+    this._options = {
+      ...defaultOptions,
+      ...options,
+    };
+  }
+
+  /**
+   * Registers the event listeners for React Native Navigation
+   */
+  public registerRoutingInstrumentation(
+    listener: TransactionCreator,
+    beforeNavigate: BeforeNavigate
+  ): void {
+    super.registerRoutingInstrumentation(listener, beforeNavigate);
+
+    this._navigation
+      .events()
+      .registerCommandListener(this._onCommand.bind(this));
+
+    this._navigation
+      .events()
+      .registerComponentWillAppearListener(
+        this._onComponentWillAppear.bind(this)
```

Comment:
Why component `componentWillAppear` over `componentDidAppear`?

Reply:
We want the earliest confirmation that the screen actually did change but I don't think it really makes a difference. If we want to add a span to measuring the time like you suggested in the other comment, I'd change this to instrument `componentDidAppear`.